### ISSUE TITLE
The Overlap Calculated by the Toolkit (region_overlap) seems different from standard IoU

### ISSUE THREAD

**Issue by huanglianghua** [votchallenge/vot-toolkit#202]
Run function `region_overlap` on two simple rectangles:

```
r1 = {[1, 1, 100, 100]};
r2 = {[50, 50, 100, 100]};
out = region_overlap(r1, r2);
```

The output is `0.1528`, which is different from standard iou as is used in, e.g., OTB or VOCdevkit, which produces `0.1495`.

Am I making any mistakes? Or the compiled region overlap function is only an approximation?

My calculation:

```
area1 = 100 * 100;
area2 = 100 * 100;
area_inter = 51 * 51;
overlap = area_inter / (area1 + area2 - area_inter);
```

The output is 0.1495.

**Comment by huanglianghua** [comment]
Besides the inaccurate IoU calculation, it seems that the code contains so many bugs.

For example, in function `estimate_expected_overlap.m`, around line `101`:

```MATLAB
...
for j = 1:numel(points)-1;
    o = segments{i}(points(j):points(j+1)); o(isnan(o)) = 0;
    fragments(f, :) = 0;
    fragments(f, 1:min(numel(o), fragments_length)) = o;
...
```

`fragments(f, :)` should have been `nan` instead of `0`, so that the frames outside video frame range wouldn't have impact on EAO calculation.

This makes the calculation of expected average overlaps incorrect.

**Comment by yl1991** [comment]
@lukacu

**Comment by lukacu** [comment]
I will have a look at it, the process of overlap calculation is a bit more complex than in OTB, because we are rasterizing all regions.

**Comment by huanglianghua** [comment]
Hi @lukacu, as a reference, I had tried using the Polygon intersection (`shapely.geometry.Polygon.intersection` of Python package `shapely`) for the calculation of IoU between rotated rectangles. The result is identical with the OTB/VOCdevkit outputs (0.1495 in the above case) when the regions are rectangles, while they differs from the output of `region_overlap` in this repo.

**Comment by lukacu** [comment]
Ok, so I have checked the code and there was a kind of bug in the rectangle to polygon conversion in MEX wrapper for the overlap calculation library, every rectangle was 1px wider than it should be. Actually the C code had this bug long ago but was fixed. The C code was ok and was correctly reporting 0.1495, but the MEX was doing it incorrectly if you have specified axis-aligned rectangles. I will commit the fix now, closing issue.
